File: modules/agent/src/controllers/contact_controller.py
"""Controller layer for contact routing to services."""

from typing import Optional, List

from fastapi import Request

from lib.decorators import handle_http_exceptions
from serializers.common import to_paged_response
from serializers.contact import contact_to_list_response, contact_to_detail_response
from schemas.contact import ContactCreate, ContactUpdate
from services.contact_service import (
    get_contacts_paginated,
    search_contacts as search_contacts_service,
    get_contact as get_contact_service,
    create_contact as create_contact_service,
    update_contact as update_contact_service,
    delete_contact as delete_contact_service,
)

# Re-export for routes
__all__ = ["ContactCreate", "ContactUpdate"]


@handle_http_exceptions
async def get_contacts(
    page: int,
    limit: int,
    order_by: str,
    order: str,
    search: Optional[str] = None,
) -> dict:
    """Get contacts with pagination."""
    contacts, total = await get_contacts_paginated(page, limit, order_by, order, search)
    items = [contact_to_list_response(c) for c in contacts]
    return to_paged_response(total, page, limit, items)


@handle_http_exceptions
async def search_contacts(request: Request, query: str) -> List[dict]:
    """Search contacts and individuals."""
    tenant_id = request.state.tenant_id
    return await search_contacts_service(query, tenant_id)


@handle_http_exceptions
async def get_contact(contact_id: int) -> dict:
    """Get contact by ID."""
    contact = await get_contact_service(contact_id)
    return contact_to_detail_response(contact)


@handle_http_exceptions
async def create_contact(request: Request, data: ContactCreate) -> dict:
    """Create a new contact."""
    user_id = request.state.user_id
    contact_data = data.model_dump(exclude={"account_ids"})
    contact = await create_contact_service(contact_data, user_id, data.account_ids)
    return contact_to_detail_response(contact)


@handle_http_exceptions
async def update_contact(request: Request, contact_id: int, data: ContactUpdate) -> dict:
    """Update a contact."""
    user_id = request.state.user_id
    contact_data = data.model_dump(exclude={"account_ids"}, exclude_unset=True)
    contact = await update_contact_service(contact_id, contact_data, user_id, data.account_ids)
    return contact_to_detail_response(contact)


@handle_http_exceptions
async def delete_contact(contact_id: int) -> dict:
    """Delete a contact."""
    await delete_contact_service(contact_id)
    return {"success": True}

File: modules/agent/internal/controllers/contact_controller.go
package controllers

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"agent/internal/middleware"
	"agent/internal/schemas"
	"agent/internal/serializers"
	"agent/internal/services"
)

// ContactController handles contact HTTP requests
type ContactController struct {
	contactService *services.ContactService
}

// NewContactController creates a new contact controller
func NewContactController(contactService *services.ContactService) *ContactController {
	return &ContactController{contactService: contactService}
}

// GetContacts handles GET /contacts
func (c *ContactController) GetContacts(w http.ResponseWriter, r *http.Request) {
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	orderBy := r.URL.Query().Get("orderBy")
	order := r.URL.Query().Get("order")
	search := r.URL.Query().Get("search")

	var tenantID *int64
	if tid, ok := middleware.GetTenantID(r.Context()); ok {
		tenantID = &tid
	}

	result, err := c.contactService.GetContacts(page, limit, orderBy, order, search, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, result)
}

// GetContact handles GET /contacts/{id}
func (c *ContactController) GetContact(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid contact ID")
		return
	}

	result, err := c.contactService.GetContact(id)
	writeContactResponse(w, http.StatusOK, result, err)
}

// CreateContact handles POST /contacts
func (c *ContactController) CreateContact(w http.ResponseWriter, r *http.Request) {
	var input schemas.ContactCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	userID, _ := middleware.GetUserID(r.Context())

	result, err := c.contactService.CreateContact(input, userID)
	writeContactResponse(w, http.StatusCreated, result, err)
}

// UpdateContact handles PUT /contacts/{id}
func (c *ContactController) UpdateContact(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid contact ID")
		return
	}

	var input schemas.ContactUpdate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	userID, _ := middleware.GetUserID(r.Context())

	result, err := c.contactService.UpdateContact(id, input, userID)
	writeContactResponse(w, http.StatusOK, result, err)
}

// DeleteContact handles DELETE /contacts/{id}
func (c *ContactController) DeleteContact(w http.ResponseWriter, r *http.Request) {
	idStr := chi.URLParam(r, "id")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid contact ID")
		return
	}

	err = c.contactService.DeleteContact(id)
	if errors.Is(err, services.ErrContactNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, serializers.SuccessResponse{Success: true})
}

// SearchContacts handles GET /contacts/search
func (c *ContactController) SearchContacts(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("q")
	if query == "" {
		writeError(w, http.StatusBadRequest, "search query is required")
		return
	}

	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))

	results, err := c.contactService.SearchContacts(query, limit)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, results)
}

func writeContactResponse(w http.ResponseWriter, successStatus int, result *serializers.ContactResponse, err error) {
	if errors.Is(err, services.ErrContactNotFound) {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	if errors.Is(err, services.ErrContactNameRequired) {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, successStatus, result)
}
